From 227ede8af397a30035c66cf608e80f048abe708a Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 4 Jan 2019 15:09:50 +0800
Subject: [PATCH 4/5] pds: non-root user can switch task among
 ISO/NORMAL/BATCH/IDLE.

non-root users now can switch task among ISO/NORMAL/BATCH/IDLE policy
freely.
---
 kernel/sched/pds.c | 21 +--------------------
 1 file changed, 1 insertion(+), 20 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 8f821622084f..a4fb93dfd8ca 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -4315,7 +4315,7 @@ __sched_setscheduler(struct task_struct *p,
 	 * Allow unprivileged RT tasks to decrease priority:
 	 */
 	if (user && !capable(CAP_SYS_NICE)) {
-		if (rt_policy(policy)) {
+		if (SCHED_FIFO == policy || SCHED_RR == policy) {
 			unsigned long rlim_rtprio =
 					task_rlimit(p, RLIMIT_RTPRIO);
 
@@ -4327,25 +4327,6 @@ __sched_setscheduler(struct task_struct *p,
 			if (attr->sched_priority > p->rt_priority &&
 			    attr->sched_priority > rlim_rtprio)
 				return -EPERM;
-		} else {
-			switch (p->policy) {
-				/*
-				 * Can only downgrade policies but not back to
-				 * SCHED_NORMAL
-				 */
-				case SCHED_BATCH:
-					if (policy == SCHED_BATCH)
-						return 0;
-					if (policy != SCHED_IDLE)
-						return -EPERM;
-					break;
-				case SCHED_IDLE:
-					if (policy == SCHED_IDLE)
-						return 0;
-					return -EPERM;
-				default:
-					break;
-			}
 		}
 
 		/* Can't change other user's priorities */
-- 
2.20.1.98.gecbdaf0899

